third_party/airflow/armada/operators/armada_deferrable.py

```diff
@@ -0,0 +1,315 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import logging
+from typing import Optional, Sequence, Tuple, Any
+
+import grpc
+
+from airflow.exceptions import AirflowException
+from airflow.models import BaseOperator
+from airflow.triggers.base import BaseTrigger, TriggerEvent
+
+from armada_client.client import ArmadaClient
+from armada.operators.jobservice import JobServiceClient
+from armada.operators.jobservice_asyncio import JobServiceAsyncIOClient
+
+from armada.operators.utils import (
+    airflow_error,
+    search_for_job_complete_async,
+    annotate_job_request_items,
+)
+from armada.jobservice import jobservice_pb2
+
+
+armada_logger = logging.getLogger("airflow.task")
+
+
+class ArmadaDeferrableOperator(BaseOperator):
+    """
+    Implementation of a deferrable armada operator for airflow.
+
+    Distinguished from ArmadaOperator by its ability to defer itself after
+    submitting its job_request_items.
+
+    See https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/deferring.html
+    for more information about deferrable airflow operators.
+
+    Airflow operators inherit from BaseOperator.
+
+    :param name: The name of the airflow task.
+    :param armada_channel_args: GRPC channel arguments to be used when creating
+      a grpc channel to connect to the armada server instance.
+    :param job_service_channel_args: GRPC channel arguments to be used when creating
+      a grpc channel to connect to the job service instance.
+    :param armada_queue: The queue name for Armada.
+    :param job_request_items: A PodSpec that is used by Armada for submitting a job.
+    :param lookout_url_template: A URL template to be used to provide users
+        a valid link to the related lookout job in this operator's log.
+        The format should be:
+        "https://lookout.armada.domain/jobs?job_id=<job_id>" where <job_id> will
+        be replaced with the actual job ID.
+
+    :return: A deferrable armada operator instance.
+    """  # noqa
+
+    def __init__(
+        self,
+        name: str,
+        armada_channel_args: dict,
+        job_service_channel_args: dict,
+        armada_queue: str,
+        job_request_items,
```

Comment:
Can you add a type here?  

It would also be ideal to have the shape of the armada_channel and job_service_channel_args.  dict is basically a nontype.